pkg/cloudprovider/provider/baremetal/plugins/tinkerbell/driver.go

```diff
@@ -18,205 +18,193 @@ package tinkerbell
-	TinkServerAddress string
-	ImageRepoAddress  string
+	ClusterName string
+	OSImageURL  string
+	TinkClient  ctrlruntimeclient.Client
+	//KubeClient     ctrlruntimeclient.Client
+	HardwareRef    types.NamespacedName
+	HardwareClient client.HardwareClient
+	WorkflowClient client.WorkflowClient
+	TemplateClient client.TemplateClient
+}
 
-	metadataClient metadataclient.Client
-	hardwareClient tinkerbellclient.HardwareClient
-	templateClient tinkerbellclient.TemplateClient
-	workflowClient tinkerbellclient.WorkflowClient
+func init() {
+	// Ensure the Tinkerbell API types are registered with the global scheme.
+	if err := tinkv1alpha1.SchemeBuilder.AddToScheme(scheme.Scheme); err != nil {
+		panic(fmt.Sprintf("failed to add tinkv1alpha1 to scheme: %v", err))
+	}
 }
 
 // NewTinkerbellDriver returns a new TinkerBell driver with a configured tinkserver address and a client timeout.
-func NewTinkerbellDriver(mdConfig *metadataclient.Config, factory ClientFactory, tinkServerAddress, imageRepoAddress string) (plugins.PluginDriver, error) {
-	if tinkServerAddress == "" || imageRepoAddress == "" {
-		return nil, errors.New("tink-server address, ImageRepoAddress cannot be empty")
+func NewTinkerbellDriver(tinkConfig tinktypes.Config, tinkSpec *tinktypes.TinkerbellPluginSpec) (plugins.PluginDriver, error) {
+	tinkClient, err := ctrlruntimeclient.New(tinkConfig.RestConfig, ctrlruntimeclient.Options{Scheme: scheme.Scheme})
+	if err != nil {
+		return nil, fmt.Errorf("failed to create k8s client: %w", err)
 	}
 
-	var (
-		mdClient   metadataclient.Client
-		hwClient   tinkerbellclient.HardwareClient
-		tmplClient tinkerbellclient.TemplateClient
-		wflClient  tinkerbellclient.WorkflowClient
-		err        error
-	)
+	if err != nil {
+		return nil, err
+	}
 
-	if factory == nil {
-		mdClient, err = metadataclient.NewMetadataClient(mdConfig)
-		if err != nil {
-			return nil, fmt.Errorf("failed to create metadata client: %w", err)
-		}
+	hwClient := client.NewHardwareClient(tinkClient)
 
-		if err := tinkclient.Setup(); err != nil {
-			return nil, fmt.Errorf("failed to setup tink-server client: %w", err)
-		}
+	wkClient := client.NewWorkflowClient(tinkClient)
 
-		hwClient = tinkerbellclient.NewHardwareClient(tinkclient.HardwareClient)
-		tmplClient = tinkerbellclient.NewTemplateClient(tinkclient.TemplateClient)
-		wflClient = tinkerbellclient.NewWorkflowClient(tinkclient.WorkflowClient, tinkerbellclient.NewHardwareClient(tinkclient.HardwareClient))
-	} else {
-		mdClient, hwClient, tmplClient, wflClient = factory()
-	}
+	tmplClient := client.NewTemplateClient(tinkClient)
 
-	d := &driver{
-		TinkServerAddress: tinkServerAddress,
-		ImageRepoAddress:  imageRepoAddress,
-		metadataClient:    mdClient,
-		hardwareClient:    hwClient,
-		templateClient:    tmplClient,
-		workflowClient:    wflClient,
+	d := driver{
+		ClusterName: tinkSpec.ClusterName.Value,
+		TinkClient:  tinkClient,
+		HardwareRef: tinkSpec.HardwareRef,
+		//KubeClient:     k8sClient,
```

Comment:
Another leftover

Reply:
Probably yeah, it is from the original PR. I just wanted to know what's the idea of this field. But this was at the beginning of my implementation, but since I have refactored quite a lot I can simply take it out.